sw/device/mask_rom/boot.md

```diff
@@ -1,71 +1,134 @@
 
 2. Execution Begins
-  - CRT Startup Code (ASM)
-    - Clean Slate (SRAM, Registers)
-    - Well-defined interrupt/exception handler
-        Open Q: Do we want to enable interrupts in mask rom?
-    - Jump into C code (`boot`):
+   *   CRT Startup Code (Written in Assembly)
+
+       *   Clean Device State Part 1. (Registers, main SRAM)
+
+       *   Well-defined interrupt/exception handler.
+
+           **Open Q:** Do we want to enable interrupts in mask rom?
+
+       *   Setup structures for C execution (`.data` & `.bss` sections, stack).
+
+       *   Jump into C code.
+
+   *   Main Secure Boot Software (Written in C)
+       *   Orchestrated from `boot`:
 
 ```c
 void boot(void) {
-  // - Clean all other hardware state (using DIF Reset functions)
+  // - Clean Device State Part 2
   clean_device_state();
 
+  // - Enable Memory Protection
+  //   - SRAM Scramble
+  //   - PMP Initial Region (if not done in power on)
+  enable_memory_protection();
+
   // - Chip-specific startup functionality
+  //   **Open Q:** Proprietary Software Strategy.
   //   - Clocks
   //   - AST
   //   - Flash
   //   - Entropy
-  //   Open Q: Proprietary Software Strategy.
   chip_specific_startup();
 
   // - Read Boot Policy for ROM_EXT from Flash (2.b)
-  //   Open Q: How is this protected beyond being in flash?
+  //   **Open Q:** How is this protected beyond being in flash?
   boot_policy = read_boot_policy();
 
-  // - Determine which rom_ext slot is in use (2.b,2.c.i)
-  rom_ext_manifest = read_rom_ext_slot(boot_policy);
-
-  // Verify ROM_EXT Image Digest (2.c.ii)
-  if (!verify_rom_ext_hash(rom_ext_manifest)) {
-    // Boot Failure (check policy)
+  // - Determine which ROM_EXT slot is prioritised (2.b, 2.c.i)
+  for ( current_rom_ext_manifest in rom_ext_manifests_to_try(boot_policy) ) {
+
+    // Verify ROM_EXT Image Digest (2.c.ii)
+    // **Open Q:** Integration with Secure Boot Hardware
+    if (!verify_rom_ext_hash(current_rom_ext_manifest)) {
+      // Boot Failure (check policy)
+      if (try_next_on_digest_failed(boot_policy))
+        continue
+      else
+        break
+    }
+
+    // Verify ROM_EXT Image Signature (2.c.iii)
+    // **Open Q:** Integration with Secure Boot Hardware
+    valid, signature = verify_rom_ext_signature(current_rom_ext_manifest);
+    if (!valid) {
+      // Boot Failure (check policy)
+      if (try_next_on_signature_failed(boot_policy))
+        continue
+      else
+        break
+    }
+
+    // System State Measurements (2.c.iv)
+    measurements = perform_system_state_measurements();
+    if (!boot_allowed(measurements)) {
+      // Lifecycle failure (no policy check)
+      break
+    }
+
+    // CreatorRootKey (2.c.iv)
+    root_key_identity = derive_creator_root_key(measurements);
+    load_root_key(root_key_identity);
+
+    // PMP Region for ROM_EXT (2.c.v)
+    // **Open Q:** Integration with Secure Boot Hardware
+    // **Open Q:** Do we need to lock down Boot ROM at final jump?
```

Comment:
Is it even possible to lock down boot rom from boot rom?

Reply:
This has been an open question for a while. We originally wanted to write a trampoline into SRAM that did the final PMP lock and then jump. However, now that SRAM is non-executable we need a different mechanism. We need to finalize the details surrounding the sensitivity of the mask ROM itself. We may be able to get away with just locking out the proprietary regions before jumping away. Or placing the final lock and jump routine just past a page boundary (this wastes ROM space though).